src/kakarot/account.cairo

```diff
@@ -0,0 +1,337 @@
+        let (storage_start) = default_dict_new(0);
+        return new model.Account(
+            address=address,
+            code_len=code_len,
+            code=code,
+            storage_start=storage_start,
+            storage=storage_start,
+            nonce=nonce,
+            selfdestruct=0,
+        );
+    }
+
+    // @dev Copy the Account to safely mutate the storage
+    // @param self The pointer to the Account
+    func copy{range_check_ptr}(self: model.Account*) -> model.Account* {
+        let (storage_start, storage) = default_dict_copy(self.storage_start, self.storage);
+        return new model.Account(
+            address=self.address,
+            code_len=self.code_len,
+            code=self.code,
+            storage_start=storage_start,
+            storage=storage,
+            nonce=self.nonce,
+            selfdestruct=self.selfdestruct,
+        );
+    }
+
+    // @dev Squash dicts used internally
+    // @param self The pointer to the Account
+    // @return a Summary Account, frozen
+    func finalize{range_check_ptr}(self: model.Account*) -> Summary* {
+        let (storage_start, storage) = default_dict_finalize(self.storage_start, self.storage, 0);
+        return new Summary(
+            address=self.address,
+            code_len=self.code_len,
+            code=self.code,
+            storage_start=storage_start,
+            storage=storage,
+            nonce=self.nonce,
+            selfdestruct=self.selfdestruct,
+        );
+    }
+
+    // @notice Commit the account to the storage backend at given address
+    // @dev Account is deployed here if it doesn't exist already
+    // @dev Works on Account.Summary to make sure only finalized accounts are committed.
+    // @param self The pointer to the Account
+    // @param starknet_address A starknet address to commit to
+    // @notice Iterate through the storage dict and update the Starknet storage
+    func commit{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        self: Summary*, starknet_address: felt
+    ) {
+        alloc_locals;
+
+        let (registered_starknet_account) = Accounts.get_starknet_address(self.address);
+        let starknet_account_exists = is_not_zero(registered_starknet_account);
+
+        // Case new Account
+        if (starknet_account_exists == 0) {
+            // Deploy accounts
+            let (class_hash) = contract_account_class_hash.read();
+            Accounts.create(class_hash, self.address);
+            // Write bytecode
+            IContractAccount.write_bytecode(starknet_address, self.code_len, self.code);
+            // Set nonce
+            IContractAccount.set_nonce(starknet_address, self.nonce);
+            // Save storages
+            Internals._save_storage(starknet_address, self.storage_start, self.storage);
+            return ();
+        }
+
+        // Case SELFDESTRUCT
+        if (self.selfdestruct != 0) {
```

Comment:
in case of selfdestruct, we write the bytecode and then erase it right after?

Reply:
where do we write? There is probably a non handled edge case here when a newly deployed contrat called SELFDESTRUCT

Reply:
yes thx that's what I meant, I guess it still works but we use a lot of gas?